File: ruby/grep_std.rb
# frozen_string_literal: true
# rubygrep_std - idiomatic single-threaded Ruby: recursive walk + whole-file
# binread + String#index literal scan + NUL-in-first-64KB binary skip.
# Mirrors python/grep_std.py. Strings are forced to ASCII-8BIT so index/rindex
# work in byte positions (C-backed); ASCII case-fold is a tr (also C). No regex.

NL = "\n".b
NUL = "\x00".b

def search_file(path, pat, lpat, ci, multi, out)
  data = begin
    File.binread(path)
  rescue SystemCallError
    return false
  end
  return false if data.empty?
  n = data.bytesize
  peek = n > 65536 ? data.byteslice(0, 65536) : data
  return false if peek.include?(NUL)            # binary skip

  if ci
    hay = data.tr('A-Z', 'a-z')
    needle = lpat
  else
    hay = data
    needle = pat
  end

  matched = false
  pos = 0
  while pos <= n
    m = hay.index(needle, pos)
    break if m.nil?
    # phantom empty line after a trailing newline: grep emits nothing there
    break if m == n && n.positive? && data.getbyte(n - 1) == 0x0A
    ls = m.zero? ? 0 : (data.rindex(NL, m - 1) || -1) + 1
    le = data.index(NL, m) || n
    matched = true
    out << path << ':' if multi
    out << data.byteslice(ls, le - ls) << "\n"
    pos = le + 1
  end
  matched
end

def walk(dir, pat, lpat, ci, multi, out)
  matched = false
  entries = begin
    Dir.children(dir)
  rescue SystemCallError
    return false
  end
  entries.each do |e|
    p = File.join(dir, e)
    begin
      st = File.lstat(p)
    rescue SystemCallError
      next
    end
    next if st.symlink?                          # don't follow symlinks (grep -r)
    if st.directory?
      matched = true if walk(p, pat, lpat, ci, multi, out)
    elsif st.file?
      matched = true if search_file(p, pat, lpat, ci, multi, out)
    end
  end
  matched
end

def main(argv)
  ci = r = false
  no_more = false
  pat = nil
  paths = []
  argv.each do |a|
    if !no_more && a.start_with?('-') && a != '-'
      if a == '--'
        no_more = true
        next
      end
      a[1..].each_char do |q|
        case q
        when 'i' then ci = true
        when 'r' then r = true
        else
          warn 'usage: rubygrep_std [-r] [-i] PATTERN PATH...'
          return 2
        end
      end
    elsif pat.nil?
      pat = a
    else
      paths << a
    end
  end
  if pat.nil? || paths.empty?
    warn 'usage: rubygrep_std [-r] [-i] PATTERN PATH...'
    return 2
  end

  patb = pat.b
  lpat = patb.tr('A-Z', 'a-z')
  multi = r || paths.size > 1

  out = +''.b
  matched = false
  paths.each do |p|
    st = begin
      File.stat(p)                               # follow symlinks at top level
    rescue SystemCallError
      next
    end
    if st.directory?
      matched = true if r && walk(p, patb, lpat, ci, multi, out)
    elsif st.file?
      matched = true if search_file(p, patb, lpat, ci, multi, out)
    end
  end

  STDOUT.binmode
  STDOUT.write(out) unless out.empty?
  matched ? 0 : 1
end

exit main(ARGV)

File: perl/grep_std.pl
#!/usr/bin/env perl
# perlgrep_std - idiomatic single-threaded Perl: recursive readdir walk +
# whole-file slurp + index() literal scan + NUL-in-first-64KB binary skip.
# Mirrors python/grep_std.py. index()/rindex are C-backed under the
# interpreter; ASCII case-fold is a tr/// (also C). No regex (this project's
# literal, regex-free semantics == grep -F).
use strict;
use warnings;

# search one file, appending matching lines to $$out; returns 1 if any matched.
sub search_file {
    my ($path, $pat, $lpat, $ci, $multi, $out) = @_;
    open(my $fh, '<:raw', $path) or return 0;
    my $data = do { local $/; <$fh> };
    close($fh);
    return 0 unless defined $data;
    my $n = length $data;
    return 0 if $n == 0;
    my $peek = $n > 65536 ? substr($data, 0, 65536) : $data;
    return 0 if index($peek, "\0") >= 0;          # binary skip

    my ($hay, $needle);
    if ($ci) { ($hay = $data) =~ tr/A-Z/a-z/; $needle = $lpat; }
    else     { $hay = $data;                  $needle = $pat;  }

    my $matched = 0;
    my $pos = 0;
    while ($pos <= $n) {
        my $m = index($hay, $needle, $pos);
        last if $m < 0;
        # phantom empty line after a trailing newline: grep emits nothing there
        last if $m == $n && $n > 0 && substr($data, $n - 1, 1) eq "\n";
        my $ls = $m == 0 ? 0 : rindex($data, "\n", $m - 1) + 1;  # prev '\n'+1 (or 0)
        my $le = index($data, "\n", $m);
        $le = $n if $le < 0;
        $matched = 1;
        $$out .= "$path:" if $multi;
        $$out .= substr($data, $ls, $le - $ls) . "\n";
        $pos = $le + 1;
    }
    return $matched;
}

# recursive walk: skip symlinks (grep -r), recurse dirs, search regular files.
sub walk {
    my ($dir, $pat, $lpat, $ci, $multi, $out) = @_;
    my $matched = 0;
    opendir(my $dh, $dir) or return 0;
    my @ents = readdir $dh;
    closedir $dh;
    for my $e (@ents) {
        next if $e eq '.' || $e eq '..';
        my $p = "$dir/$e";
        next if -l $p;                            # don't follow symlinks
        if    (-d _) { $matched = 1 if walk($p, $pat, $lpat, $ci, $multi, $out); }
        elsif (-f _) { $matched = 1 if search_file($p, $pat, $lpat, $ci, $multi, $out); }
    }
    return $matched;
}

sub main {
    my ($ci, $r, $no_more) = (0, 0, 0);
    my $pat;
    my @paths;
    for my $a (@ARGV) {
        if (!$no_more && $a =~ /^-/ && $a ne '-') {
            if ($a eq '--') { $no_more = 1; next; }
            for my $q (split //, substr($a, 1)) {
                if    ($q eq 'i') { $ci = 1; }
                elsif ($q eq 'r') { $r  = 1; }
                else { print STDERR "usage: perlgrep_std [-r] [-i] PATTERN PATH...\n"; return 2; }
            }
        } elsif (!defined $pat) { $pat = $a; }
        else                    { push @paths, $a; }
    }
    if (!defined $pat || !@paths) {
        print STDERR "usage: perlgrep_std [-r] [-i] PATTERN PATH...\n";
        return 2;
    }

    my $lpat = $pat;
    $lpat =~ tr/A-Z/a-z/;
    my $multi = $r || @paths > 1;

    my $out = '';
    my $matched = 0;
    for my $p (@paths) {
        my @st = stat $p;                         # follow symlinks at top level
        next unless @st;
        if    (-d _) { $matched = 1 if $r && walk($p, $pat, $lpat, $ci, $multi, \$out); }
        elsif (-f _) { $matched = 1 if search_file($p, $pat, $lpat, $ci, $multi, \$out); }
    }

    binmode STDOUT;
    print STDOUT $out if length $out;
    return $matched ? 0 : 1;
}

exit main();
